common/metrics/events_scope.go

```diff
@@ -0,0 +1,173 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"time"
+
+	"golang.org/x/exp/event"
+	"golang.org/x/exp/maps"
+)
+
+type (
+	eventsScope struct {
+		provider   MetricProvider
+		serviceIdx ServiceIdx
+		scopeId    int
+		scopeTags  []Tag
+		scopeDef   scopeDefinition
+	}
+
+	eventsStopwatch struct {
+		recordFunc timerRecordFunc
+		start      time.Time
+	}
+
+	timerRecordFunc func(time.Duration)
+)
+
+var (
+	_ Scope     = (*eventsScope)(nil)
+	_ Stopwatch = (*eventsStopwatch)(nil)
+)
+
+func newEventsScope(provider MetricProvider, idx ServiceIdx, id int) eventsScope {
+	return eventsScope{
+		provider:   provider,
+		serviceIdx: idx,
+		scopeId:    id,
+		scopeTags:  append(make([]Tag, 0), scopeDefToTags(ScopeDefs[idx][id])...),
```

Comment:
use append(nil, ...)

Reply:
not nil but `[]Tag{}`?

Reply:
nil is better here as []Tag{} still consume something which is no use.